### PR TITLE
Support PD Auto-scaling API in pdapi

### PR DESCRIPTION
**Pull request by howardlau1999**
### What problem does this PR solve? 

Related https://github.com/pingcap/tidb-operator/issues/3159

### What is changed and how does it work?

Add a method `GetAutoscalingPlans` in pdapi

### Check List 

Tests 

 - Unit test

Code changes

 - Has Go code change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: pkg/util/http/httputil.go [hunk 1]
```diff
@@ -89,3 +89,25 @@ func DoBodyOK(httpClient *http.Client, apiURL, method string) ([]byte, error) {
...
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode >= 400 {
+		errMsg := fmt.Errorf("Error response %v URL %s,body response: %s", res.StatusCode, apiURL, string(body[:]))
+		return nil, errMsg
+	}
+	return body, err
+}
```
Comment (lichunzhu) [inline pkg/util/http/httputil.go:113]: Can we just reuse `DoBodyOK`, just like `DeleteBodyOK` did?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: pkg/pdapi/pdapi.go [hunk 2]
```diff
@@ -186,6 +189,52 @@ type MembersInfo struct {
...
+// Resource represents a kind of resource set including CPU, memory, storage.
+type Resource struct {
+	ResourceType string `json:"resource_type"`
+	// The basic unit of CPU is milli-core.
+	CPU uint64 `json:"cpu"`
+	// The basic unit of memory is byte.
+	Memory uint64 `json:"memory"`
+	// The basic unit of storage is byte.
+	Storage uint64 `json:"storage"`
+	Count   uint64 `json:"count"`
```
Comment (DanielZhangQD) [inline pkg/pdapi/pdapi.go:227]:
```suggestion
	Count   *uint64 `json:"count,omitempty"`
```

File: pkg/pdapi/pdapi.go [hunk 3]
```diff
@@ -186,6 +189,52 @@ type MembersInfo struct {
...
+// Rule is a set of constraints for a kind of component.
+type Rule struct {
+	Component   string       `json:"component"`
+	CPURule     *CPURule     `json:"cpu_rule,omitempty"`
+	StorageRule *StorageRule `json:"storage_rule,omitempty"`
+}
+
+// CPURule is the constraints about CPU.
+type CPURule struct {
+	MaxThreshold  float64  `json:"max_threshold"`
```
Comment (DanielZhangQD) [inline pkg/pdapi/pdapi.go:207]: Please help update the field name in https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go#L108-L114 to match with the names here.

File: pkg/pdapi/pdapi.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/pdapi/pdapi.go:227]: If `count` is not set, it indicates no limit.

File: pkg/pdapi/pdapi.go [hunk 4]
```diff
@@ -621,6 +670,20 @@ func (pc *pdClient) TransferPDLeader(memberName string) error {
 	return fmt.Errorf("failed %v to transfer pd leader to %s,error: %v", res.StatusCode, memberName, err2)
 }
 
+func (pc *pdClient) GetAutoscalingPlans(strategy Strategy) ([]Plan, error) {
+	apiURL := fmt.Sprintf("%s/%s", pc.url, autoscalingPrefix)
+	body, err := httputil.PostBodyOK(pc.httpClient, apiURL)
```
Comment (DanielZhangQD) [inline pkg/pdapi/pdapi.go:675]: No need to set input here?

**Comment by howardlau1999** [comment]
@DanielZhangQD PTAL

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 5]
```diff
@@ -106,12 +106,12 @@ type AutoResource struct {
...
 // AutoRule describes the rules for auto-scaling with PD API
 type AutoRule struct {
-	// Max defines the threshold to scale out
-	Max *float64 `json:"max,omitempty"`
-	// Min defines the threshold to scale in
-	Min *float64 `json:"min,omitempty"`
-	// ResourceType defines the resource types that can be used for scaling
-	ResourceType []string `json:"resource_type,omitempty"`
+	// MaxThreshold defines the threshold to scale out
+	MaxThreshold *float64 `json:"max_threshold,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:110]:
```suggestion
	MaxThreshold float64 `json:"max_threshold"`
```

Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:110]: This should be required.

File: pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go [hunk 6]
```diff
@@ -106,12 +106,12 @@ type AutoResource struct {
...
+	// MinThreshold defines the threshold to scale in
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go:111]:
```suggestion
	// MinThreshold defines the threshold to scale in, not applicable to `storage` rule
```

File: pkg/pdapi/pdapi.go [hunk 7]
```diff
@@ -621,6 +673,24 @@ func (pc *pdClient) TransferPDLeader(memberName string) error {
 	return fmt.Errorf("failed %v to transfer pd leader to %s,error: %v", res.StatusCode, memberName, err2)
 }
 
+func (pc *pdClient) GetAutoscalingPlans(strategy Strategy) ([]Plan, error) {
```
Comment (DanielZhangQD) [inline pkg/pdapi/pdapi.go:676]: The construction of `strategy` will be done in another PR?

Comment (howardlau1999) [reply pkg/pdapi/pdapi.go:676]: yes

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by rleungx — APPROVED** [review]
LGTM
